src/shared/copy.c

```diff
@@ -794,6 +797,61 @@ static int prepare_nocow(int fdf, const char *from, int fdt, unsigned *chattr_ma
         return 0;
 }
 
+/* Copies fs-verity status.  May re-open fdt to do its job. */
+static int copy_fs_verity(int fdf, int *fdt) {
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt);
+        assert(*fdt >= 0);
+
+        r = fd_verify_regular(fdf);
+        if (r < 0)
+                return r;
+
+        struct fsverity_descriptor desc = {};
+        struct fsverity_read_metadata_arg read_arg = {
+                .metadata_type = FS_VERITY_METADATA_TYPE_DESCRIPTOR,
+                .buf_ptr = (uintptr_t) &desc,
+                .length = sizeof(desc),
+        };
+
+        r = ioctl(fdf, FS_IOC_READ_VERITY_METADATA, &read_arg);
+        if (r < 0) {
+                /* These mean that fs-verity isn't enabled on the file, filesystem, or kernel.  In all of
+                 * those cases, the correct thing to do is to do nothing at all.
+                 */
+                if (errno == ENODATA || ERRNO_IS_NOT_SUPPORTED(errno))
+                        return 0;
+                return -errno;
+        }
+
+        if (r < (int) sizeof desc)
+                /* This is really quite unexpected... */
+                return -EINVAL;
+
+        /* Okay.  We're doing this now.  We need to re-open fdt as read-only because
+         * we can't enable fs-verity while writable file descriptors are outstanding.
+         */
+        r = fd_verify_regular(*fdt);
+        if (r < 0)
+                return r;
+
+        r = fd_reopen_replace(fdt, O_RDONLY | O_CLOEXEC | O_NOCTTY);
```

Comment:
why not use:

```
_cleanup_close int reopened_fd = -EBADF;
r = fd_reopen_condition(fdt, O_RDONLY|O_CLOEXEC|O_NOCTTY, O_ACCESS|O_PATH, &reopened_fd);
if (r < 0)
        return r;
if (reopened_fd >= 0)
        close_and_replace(fdt, reopened_fd);
```

This is a bit nicer since it means the whole function will become generic, i.e. can deal with O_PATH and O_RDWR fds, and will do the right thing of only reopening it in that case but not otherwise.